### PR TITLE
Job Priority

### PR DESCRIPTION
**Pull request by andrewstucki**
So, I ended up finding an implementation of brodal queues that seem to work nicely [here](https://github.com/cngkaygusuz/BrodalOkasakiHeap), refactored the interface a bit but threw the original license in the source.

I wrote a simple go program for loading data into faktory so I could compare load times and memory profiles. The queue priority structures add pretty much no memory overhead and capping the priority levels to 10, have pretty much the same performance as without the prioritization for enqueuing and processing.

Here are the runtimes for faktory without job prioritization using the script in `test/load/main.go`:

```
➜  load git:(priority) ✗ ./load push 100000
Enqueued 100000 jobs in 6.870994 seconds, rate: 14553.935072 jobs/s
➜  load git:(priority) ✗ ./load pop 100000
Processed 100000 jobs in 17.172457 seconds, rate: 5823.278658 jobs/s
```

Here are the runtimes for faktory with job prioritization:

```
➜  load git:(priority) ✗ ./load push 100000
Enqueued 100000 jobs in 6.994321 seconds, rate: 14297.312592 jobs/s
➜  load git:(priority) ✗ ./load pop 100000
Processed 100000 jobs in 16.859516 seconds, rate: 5931.368221 jobs/s
```

Here's a screenshot of the debug page with 100k jobs enqueued with faktory sans prioritization:

standard-100k

And here's with prioritization for 100k:

priority-100k

Finally, here's with the backpressure bumped up to 1 mil jobs in the queue:

priority-1mil

-----

I also started to clean up a bit of the typing around signed v. unsigned integers for things like queue size (since you should never have a negative queue size). Still more work to be done, but figure that can come in another PR. Also, it makes key generation a lot more simple to reason around since you can use `binary.BigEndian.PutUint64` instead of doing the bit shifting manually.

Let me know what you think.

### CODE REVIEW

**Comment by andrewstucki** [comment]
Closing and re-opening for re-build.

**Comment by mperham** [comment]
two things:

* priority should be stored as a single byte, not a full uint.
* there’s several points where you load and store in two steps, rather than add atomically. why?

**Comment by andrewstucki** [comment]
Oh, forgot to mention, here's the load timing for 1mil jobs (I didn't bother processing them):

```
➜  load git:(priority) ✗ ./load 1000000
Enqueued 1000000 jobs in 68.131330 seconds, rate: 14677.535223 jobs/s
```

Also, if you want I should be able to split out the `fetch`/`ack`/`fail` operations in the script if you want a bit more granular timings around operations (since this change should only affect `fetch` operations).

**Comment by mperham** [comment]
i’ll review more tomorrow, these are just quick notes.

**Comment by andrewstucki** [comment]
@mperham 

1) I can change the type for priority (I'm assuming since you want it capped between 0 and 10 and to store space in rocks?)
2) I initially thought that it was the limitations of `atomic` operations and unsigned numbers, until I just found [this](https://golang.org/src/sync/atomic/doc.go?#L104)--I can fix that too.

**Comment by mperham** [comment]
yeah, with a million jobs, every byte counts.

**Comment by mperham** [comment]
Do you think the Web UI should display the priority for each job in the table listing?  What about on the retry or scheduled details page for a job?

Are you expecting 1 to be high priority or 9?  Looks like 9, which makes sense.  Here's what RabbitMQ says about [priority queues for their impl](https://www.rabbitmq.com/priority.html) and they use the same style: larger == higher priority.

**Comment by andrewstucki** [comment]
@mperham yep, the way I was thinking of priority was 9 = highest priority, 1 = lowest priority. Maybe the web UI could display priority as a first step--the one oddity I was thinking about though is that because we're inserting jobs that get processed before other jobs into the queue a job might be initially on page 1 and then all of a sudden on like page 5 because of scheduling higher priority jobs in front of it.

So eventually I'm wondering if it would make sense to display jobs at different priority levels separately. Because of the way that the prefix scanning is done this actually wouldn't be too hard to add as on optional argument for the `Page` function--but I just haven't thought through the actual UX of what that would feel like, so not quite sure what to do there.

**Comment by mperham** [comment]
IME the web ui queue inspector is 99% of the time used for debugging and repairing something.  See what's in it, find a specific "bad" job by JID or jobtype.  I can't think of any reason why I'd want to filter a queue by priority.  If priority will affect pagination and where jobs are, that's a good reason to explicitly display the priority with the type and args.

**Comment by andrewstucki** [comment]
Ok, added the priority column for scheduled jobs, retries, and enqueued jobs. I don't think that displaying priority would really be useful if a job was dead or busy, right?

**Comment by mperham** [comment]
Seems reasonable to me.  👍

**Comment by mperham** [comment]
Oh, do you have those load scripts that we can use to maintain a baseline set of performance figures?  That will be super-useful to experiment with how queues perform without RocksDB, with various RocksDB tunings, etc.

**Comment by andrewstucki** [comment]
Right [here](https://github.com/contribsys/faktory/pull/68/files#diff-796923dbb61b0e0a55732e90480449d5) (it's basically partially stolen from `go_system_test.go` in the test directory).

**Comment by mperham** [comment]
Just curious, is there a reason you use uint64 for the Priority type rather than byte?  Byte would make it more clear that priorities are not an arbitrary integer value.

**Comment by andrewstucki** [comment]
It was a bit of an initial thing--used uint64 to make the prioritization more arbitrary, but then with the cap at 10 and the key encoding stuff, that all changes. I can change it.

**Comment by andrewstucki** [comment]
Also, I think I'll just make it an `uint8` (which `byte` is aliased to in go anyway)--just makes it a bit clearer that we're working with an integer.

File: job.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ type Job struct {
 	Args  []interface{} `json:"args"`
 
 	// optional
+	Priority   uint8                  `json:"priority"`
```
Comment (mperham) [inline job.go]: Should this be omitempty also, to minimize JSON size on disk/network, since we have a default?

File: webui/scheduled.ego [hunk 2]
```diff
@@ -40,6 +40,7 @@ func ego_listScheduled(w io.Writer, req *http.Request, set storage.SortedSet, co
             <th><%= t(req, "When") %></th>
             <th><%= t(req, "Queue") %></th>
             <th><%= t(req, "Job") %></th>
+            <th><%= t(req, "Priority") %></th>
```
Comment (mperham) [inline webui/scheduled.ego]: Please add an entry at the bottom of en.yml for this new UI string, so other languages will know to add it.

File: job.go (hunk 1 above)
Comment (mperham) [reply job.go]: Perhaps all access to the job's priority should be via an accessor, Priority(), so we can always return 5 as default and minimize JSON size?

**Comment by andrewstucki** [comment]
@mperham think this should be good, let me know if there's anything else

**Pull request merged**